flow/layers/backdrop_filter_layer.cc

```diff
@@ -9,6 +9,30 @@ namespace flutter {
 BackdropFilterLayer::BackdropFilterLayer(sk_sp<SkImageFilter> filter)
     : filter_(std::move(filter)) {}
 
+#ifdef FLUTTER_ENABLE_DIFF_CONTEXT
+
+void BackdropFilterLayer::Diff(DiffContext* context, const Layer* old_layer) {
+  DiffContext::AutoSubtreeRestore subtree(context);
+  auto* prev = static_cast<const BackdropFilterLayer*>(old_layer);
+  if (!context->IsSubtreeDirty()) {
+    FML_DCHECK(prev);
+    if (filter_ != prev->filter_) {
+      context->MarkSubtreeDirty(context->GetOldLayerPaintRegion(old_layer));
+    }
+  }
+
+  // Backdrop filter paints everywhere in cull rect
+  auto paint_bounds = filter_->computeFastBounds(context->GetCullRect());
```

Comment:
Should this use `SkImageFilter->filterBounds(..., ReverseDirection, ...)`?

Reply:
Possibly. At least in `ImageFilterLayer` `computeFastBounds` seems wrong.

One thing I don't quite understand, the documentation for `filterBounds` says it maps device-space rect, but ImageFilterLayer seems to apply it on local coordinate rect?

Reply:
It looks like `computeFastBounds` would be a NOP in this case, just returning the cull_rect. It apparently just does a union of the input bounds without considering how the filter will modify them. The bounds you pass in is the bounds to assume for any null inputs (null inputs represent source data that will be supplied later when you perform the filter).

Reply:
`computerFastBounds` works for blur (I tested it). But I can use `filterBounds` instead. However [this](https://github.com/flutter/engine/blob/master/flow/layers/image_filter_layer.cc#L26) is one thing I'm unclear about:
```cpp
void ImageFilterLayer::Preroll(PrerollContext* context,  SkMatrix& matrix) 
...
 if (filter_) {
    const SkIRect filter_input_bounds = child_bounds.roundOut();
    SkIRect filter_output_bounds =
        filter_->filterBounds(filter_input_bounds, SkMatrix::I(), SkImageFilter::kForward_MapDirection);
    child_bounds = SkRect::Make(filter_output_bounds);
}
```
`filterBounds` is supposed to map device space rects, so why is it used here on layer space rect?

Reply:
Preroll computes device space bounds.

Reply:
filterBounds maps device space rects wrt the transform you give it, which is Identity in this case.

Reply:
We may be losing some accuracy here. There may be a subtle difference between `tx(filterBounds(rect))` and `filterBounds(tx(rect))`.

Part of me is now wondering why paint bounds is in layer space. I suppose it could be so that retained layers don't change their paint bounds info if an ancestor changes its transform.

Reply:
Yeah. In most cases, (2x 3x pixel scaling) the paint bounds will be larger than they should be. In case where the layer is scaled down it may actually paint outside of paint_bounds.  Not directly related to diffing but something that we might want to fix for partial redraw because it could cause subtle issue for layers that are scaled down.

Reply:
Should be fixed [here](https://github.com/flutter/engine/pull/21824/commits/9bc439e747f69b37855b36b026e5ee80c173f2e9#diff-2b3e37b6e8eb4799ae30a9a014a551c3e13251ec7805c741ab7f5b5da04b1ea2R28).

Reply:
The link no longer seems to go anywhere, but in looking through the latest version of this code, I think the filter needs to be used with "makeWithLocalMatrix" as we do [here](https://github.com/flutter/engine/blob/df07060e4e2089cf05e22ce3a79fd9674438dcee/flow/layers/image_filter_layer.cc#L52) in image_filter_layer when we are applying the filter to a cached version.

Reply:
Fixed [here](https://github.com/flutter/engine/pull/21824/commits/72d0b75167f771af74eeb8f70cdf563a581cbb4a).